### PR TITLE
esp8266: Added bookmark support

### PR DESCRIPTION
**Pull request by neutralinsomniac**

### CODE REVIEW

File: esp8266/util.cpp [hunk 1]
```diff
@@ -35,7 +35,10 @@ setup(void)
 		switch (settings->revision) {
 		case 1:
 			settings->http_server = 0;
-			break;
+            /* intentional fallthrough */
```
Comment (jcs) [inline esp8266/util.cpp]: Tabs please, and use just "FALLTHROUGH" in all caps.

File: esp8266/util.cpp [hunk 2]
```diff
@@ -35,7 +35,10 @@ setup(void)
...
+		case 2:
+			memset(settings->bookmarks, 0, BOOKMARK_SIZE * NUM_BOOKMARKS);
+			strcpy(settings->bookmarks[0], "klud.ge");
```
Comment (jcs) [inline esp8266/util.cpp:41]: :)

File: esp8266/wifistation.ino [hunk 3]
```diff
@@ -202,32 +202,70 @@ exec_cmd(char *cmd, size_t len)
...
-		else {
-			errstr = strdup("invalid hostname");
+		switch (lcmd[3]) {
+		/* ATDT: dial a host */
+		case 't':
+			host = ohost = (char *)malloc(len);
+			if (host == NULL)
+				goto error;
+			host[0] = '\0';
+			if (sscanf(lcmd, "atdt%[^:]:%d%n", host, &port, &chars) == 2 &&
```
Comment (jcs) [inline esp8266/wifistation.ino]: This and any other re-indented lines need to be re-wrapped at 80 chars.

File: esp8266/wifistation.ino [hunk 4]
```diff
@@ -202,32 +202,70 @@ exec_cmd(char *cmd, size_t len)
...
+			}
+			break;
+		/* ATDS: dial a stored host */
+		case 's':
+			if (sscanf(lcmd, "atds%d", &index) != 1)
+				goto error;
+
+			if (index < 1 || index > NUM_BOOKMARKS) {
+				outputf("ERROR index must be between 1-%d\r\n",
+				NUM_BOOKMARKS);
```
Comment (jcs) [inline esp8266/wifistation.ino]: Please indent that line another 4 spaces.

File: esp8266/wifistation.ino [hunk 5]
```diff
@@ -282,6 +320,10 @@ exec_cmd(char *cmd, size_t len)
 			    mailstation_alive ? "yes" : "no");
 			outputf("HTTP Server:       %s\r\n",
 			    settings->http_server ? "yes" : "no");
+			for (int i = 0; i < NUM_BOOKMARKS; i++) {
+				outputf("Bookmark %d:	%s\r\n", i+1,
```
Comment (jcs) [inline esp8266/wifistation.ino]: I think if `bookmarks[i][0] == '\0'` we can just skip printing it.

File: esp8266/wifistation.ino [hunk 6]
```diff
@@ -641,6 +683,25 @@ exec_cmd(char *cmd, size_t len)
...
+				uint32_t index = 0;
+				/* AT&Zn=...: store address */
+				if (sscanf(lcmd, "at&z%u=", &index) != 1) {
+					output("ERROR invalid store command\r\n");
+					goto error;
+				}
+
+				if (index < 1 || index > NUM_BOOKMARKS) {
+					outputf("ERROR index must be between 1-%d\r\n",
+					NUM_BOOKMARKS);
```
Comment (jcs) [inline esp8266/wifistation.ino]: +4 spaces

File: esp8266/wifistation.ino [hunk 7]
```diff
@@ -641,6 +683,25 @@ exec_cmd(char *cmd, size_t len)
...
+		case 'z': {
+				uint32_t index = 0;
+				/* AT&Zn=...: store address */
+				if (sscanf(lcmd, "at&z%u=", &index) != 1) {
+					output("ERROR invalid store command\r\n");
+					goto error;
+				}
+
+				if (index < 1 || index > NUM_BOOKMARKS) {
+					outputf("ERROR index must be between 1-%d\r\n",
```
Comment (jcs) [inline esp8266/wifistation.ino]: Wrap at 80

File: esp8266/wifistation.ino [hunk 8]
```diff
@@ -641,6 +683,25 @@ exec_cmd(char *cmd, size_t len)
...
+					NUM_BOOKMARKS);
+					goto error;
+				}
+
+				memset(settings->bookmarks[index-1], 0, sizeof(settings->bookmarks[0]));
+				strncpy(settings->bookmarks[index-1], cmd + 6, sizeof(settings->bookmarks[0]) - 1);
```
Comment (jcs) [inline esp8266/wifistation.ino]: Wrap at 80

File: esp8266/wifistation.ino [hunk 9]
```diff
@@ -641,6 +683,25 @@ exec_cmd(char *cmd, size_t len)
...
+				output("OK\r\n");
+				break;
+			}
```
Comment (jcs) [inline esp8266/wifistation.ino:720]: This needs a corresponding `AT&Zn?` handler to just display a bookmark value.

File: esp8266/wifistation.h [hunk 10]
```diff
@@ -22,14 +22,14 @@
 #include <EEPROM.h>
 #include <ESP8266WiFi.h>
 
-#define WIFISTATION_VERSION	"0.4"
+#define WIFISTATION_VERSION	"0.5"
```
Comment (jcs) [inline esp8266/wifistation.h]: I may have some other things for 0.5, so please remove this for now.

**Review by jcs — CHANGES_REQUESTED** [review]

File: esp8266/wifistation.ino (hunk 3 above)
Comment (neutralinsomniac) [reply esp8266/wifistation.ino]: Is your tabstop set to 8?

Comment (jcs) [reply esp8266/wifistation.ino]: Yeah, basically OpenBSD's [`style(9)`](http://man.openbsd.org/style.9)

**Comment by jcs** [comment]
Thanks!

**Pull request merged**